File: EvoFighters/Eval.py
'''This module handles evaluating the parse trees that Parsing creates'''

from Parsing import COND, ACT, ATTR, VAL
from Utils import dmg_repr, sig_repr
import operator as Op
from random import randint

sd = None  # Set by Arena

class PerformableAction(object):
    '''Represents a concrete, comparable action that a creature intends to carry
    out'''
    def __init__(self, typ, arg):
        self.typ = typ
        self.arg = arg

    def __eq__(self, other):
        return self.typ == other.typ and self.arg == other.arg

    def __repr__(self):
        return str(self)

    def __str__(self):
        if self.typ == ACT['attack']:
            return "attack with damage type: {}".format(dmg_repr(self.arg))
        elif self.typ == ACT['defend']:
            return "defend against damage type: {}".format(dmg_repr(self.arg))
        elif self.typ == ACT['signal']:
            return "signal with the color {0}".format(sig_repr(self.arg))
        elif self.typ == ACT['use']:
            return "use an item in his inventory"
        elif self.typ == ACT['take']:
            return "take something from target"
        elif self.typ == ACT['wait']:
            return "wait"
        elif self.typ == ACT['flee']:
            return "flee the encounter"
        elif self.typ == ACT['mate']:
            return "mate with target"
        else:
            return "do unknown action: ({}, {})".format(self.typ, self.arg)


class InvalidInstructionError(Exception):
    '''Thrown whenever an invalid instruction is evaluated'''
    pass


def evaluate(me, tree):
    '''Eval takes information from the creature and a thought and returns an
    action to perform'''
    cond_typ = tree[0]
    if cond_typ == COND['always']:
        return eval_act(me, tree[1])
    elif cond_typ == COND['in_range']:
        val1 = get_val(me, tree[1])
        val2 = get_val(me, tree[2])
        check_val = get_val(me, tree[3])
        if min(val1, val2) <= check_val <= max(val1, val2):
            sd.print3('{val_repr} was between {} and {}', val1, val2, val_repr = tree[3])
            return eval_act(me, tree[4])
        else:
            sd.print3('{val_repr} was not between {} and {}', val1, val2, val_repr = tree[1])
            return eval_act(me, tree[5])
    elif COND['less_than'] <= cond_typ <= COND['not_equal_to']:
        if cond_typ == COND['less_than']:
            op_str = '<'
            op = Op.lt
        elif cond_typ == COND['greater_than']:
            op_str = '>'
            op = Op.gt
        elif cond_typ == COND['equal_to']:
            op_str = '=='
            op = Op.eq
        elif cond_typ == COND['not_equal_to']:
            op_str = '!='
            op = Op.ne
        val1 = get_val(me, tree[1])
        val2 = get_val(me, tree[2])
        if op(val1, val2):
            sd.print3('{val_repr} was {} {val_repr2}', op_str, val_repr = tree[1],
                   val_repr2 = tree[2])
            return eval_act(me, tree[3])
        else:
            sd.print3('{val_repr} was not {} {val_repr2}', op_str, val_repr = tree[1], 
                   val_repr2 = tree[2])
            return eval_act(me, tree[4])
    elif cond_typ in [COND['me_last_act'], COND['target_last_act']]:
        if cond_typ == COND['me_last_act']:
            actor = me
        else:
            actor = me.target
        act1 = eval_act(me, tree[1])
        if act1 == actor.last_action:
            sd.print3('{who.name}\'s last action was {act_repr}', who = actor, 
                   act_repr = tree[1])
            return eval_act(me, tree[2])
        else:
            sd.print3('{who.name}\'s last action was not {act_repr} ({fst} vs. {snd})',
                   who = actor, act_repr = tree[1], fst = act1, 
                   snd = actor.last_action)
            return eval_act(me, tree[3])
    else:
        raise InvalidInstructionError("Couldn't understand condition: {0}".
                                      format(cond_typ))

def get_val(me, tree):
    '''Evaluates a VAL node in a thought tree'''
    val_typ = tree[0]
    if val_typ == VAL['literal']:
        return tree[1]
    elif val_typ == VAL['random']:
        ret = randint(-1, 9)
        return ret
    elif val_typ == VAL['me']:
        return get_attr(me, tree[1])
    elif val_typ == VAL['target']:
        return get_attr(me.target, tree[1])

def get_attr(who, attr_typ):
    '''Returns the value of the attribute on "who" '''
    if attr_typ == ATTR['energy']:
        return who.energy
    elif attr_typ == ATTR['signal']:
        return who.signal
    elif attr_typ == ATTR['generation']:
        return who.generation
    elif attr_typ == ATTR['kills']:
        return who.kills
    elif attr_typ == ATTR['survived']:
        return who.survived
    elif attr_typ == ATTR['num_children']:
        return who.num_children
    elif attr_typ == ATTR['top_item']:
        return who.top_item if who.has_items else -1

def eval_act(me, tree):
    '''Returns an action suitable for performing (PerformableAction)'''
    act_typ = tree[0]
    if act_typ in (ACT['attack'], ACT['defend'], ACT['signal']):
        return PerformableAction(act_typ, tree[1])
    elif act_typ in (ACT['use'], ACT['take'],
                     ACT['wait'], ACT['flee'], ACT['mate']):
        return PerformableAction(act_typ, None)
    elif act_typ == ACT['subcondition']:
        return evaluate(me, tree[1])
    else:
        raise InvalidInstructionError("Didn't understand action: {0}"\
                                      .format(act_typ))

if __name__ == '__main__':
    from Parsing import Parser, TooMuchThinkingError
    from Creatures import Creature
    last_action = PerformableAction(ACT['wait'], None)
    for _ in xrange(1000):
        a = Creature()
        b = Creature()
        a.target = b
        a.last_action = last_action
        try:
            p = Parser(a.dna)
            last_action = evaluate(a, next(p).tree)
            str(last_action)
        except TooMuchThinkingError:
            continue
    

File: evofighters_rust/src/eval.rs
use std::fmt;
use std::cmp::{max, min, PartialEq, PartialOrd};

use dna::{ast, lex};
use creatures::Creature;
use rng::RngState;
use settings;

// PerformableAction is the result of evaluating a thought tree
#[derive(Debug, Copy, PartialEq, Eq, Clone, Serialize, Deserialize)]
pub enum PerformableAction {
    Attack(lex::DamageType),
    Defend(lex::DamageType),
    Signal(lex::Signal),
    Eat,
    Take,
    Wait,
    Flee,
    Mate,
    NoAction,
}

impl fmt::Display for PerformableAction {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            PerformableAction::Attack(dmg) => {
                write!(f, "attack with damage type: {:?}", dmg)
            }
            PerformableAction::Defend(dmg) => {
                write!(f, "defend against damage type: {:?}", dmg)
            }
            PerformableAction::Signal(sig) => {
                write!(f, "signal with the color: {:?}", sig)
            }
            PerformableAction::Eat => write!(f, "use an item in his inventory"),
            PerformableAction::Take => write!(f, "take something from target"),
            PerformableAction::Wait => write!(f, "wait"),
            PerformableAction::Flee => write!(f, "flee the encounter"),
            PerformableAction::Mate => write!(f, "mate with the target"),
            PerformableAction::NoAction => write!(f, "(no action)"),
        }
    }
}

pub fn evaluate(
    me: &Creature,
    other: &Creature,
    tree: &ast::Condition,
) -> PerformableAction {
    match *tree {
        ast::Condition::Always(ref action) => eval_action(me, other, action),
        ast::Condition::RangeCompare {
            ref value,
            ref bound_a,
            ref bound_b,
            ref affirmed,
            ref denied,
        } => {
            let a = eval_value(me, other, bound_a);
            let b = eval_value(me, other, bound_b);
            let check_val = eval_value(me, other, value);
            if min(a, b) <= check_val && check_val <= max(a, b) {
                trace!("{} was between {} and {}", check_val, a, b);
                eval_action(me, other, affirmed)
            } else {
                trace!("{} was not between {} and {}", check_val, a, b);
                eval_action(me, other, denied)
            }
        }
        ast::Condition::BinCompare {
            ref operation,
            ref lhs,
            ref rhs,
            ref affirmed,
            ref denied,
        } => {
            let op: fn(&usize, &usize) -> bool = match *operation {
                ast::BinOp::LT => PartialOrd::lt,
                ast::BinOp::GT => PartialOrd::gt,
                ast::BinOp::EQ => PartialEq::eq,
                ast::BinOp::NE => PartialEq::ne,
            };
            let evaled_lhs = eval_value(me, other, lhs);
            let evaled_rhs = eval_value(me, other, rhs);
            if op(&evaled_lhs, &evaled_rhs) {
                trace!(
                    "{:?}({}) was {} {:?}({})",
                    lhs,
                    evaled_lhs,
                    operation,
                    rhs,
                    evaled_rhs
                );
                eval_action(me, other, affirmed)
            } else {
                trace!(
                    "{:?}({}) was not {} {:?}({})",
                    lhs,
                    evaled_lhs,
                    operation,
                    rhs,
                    evaled_rhs
                );
                eval_action(me, other, denied)
            }
        }
        ast::Condition::ActionCompare {
            ref actor_type,
            ref action,
            ref affirmed,
            ref denied,
        } => {
            let (actor, actor_str) = match *actor_type {
                ast::ActorType::Me => (me, "my"),
                ast::ActorType::Other => (other, "other's"),
            };
            let my_action = eval_action(me, other, action);
            if my_action == actor.last_action {
                trace!(
                    "{}'s last action was {:?}",
                    actor_str,
                    actor.last_action
                );
                eval_action(me, other, affirmed)
            } else {
                trace!(
                    "{}'s last action was not {:?}",
                    actor_str,
                    actor.last_action
                );
                eval_action(me, other, denied)
            }
        }
    }
}

fn eval_action(
    me: &Creature,
    other: &Creature,
    action: &ast::Action,
) -> PerformableAction {
    match *action {
        ast::Action::Attack(dmg) => PerformableAction::Attack(dmg),
        ast::Action::Defend(dmg) => PerformableAction::Defend(dmg),
        ast::Action::Signal(sig) => PerformableAction::Signal(sig),
        ast::Action::Eat => PerformableAction::Eat,
        ast::Action::Take => PerformableAction::Take,
        ast::Action::Wait => PerformableAction::Wait,
        ast::Action::Flee => PerformableAction::Flee,
        ast::Action::Mate => PerformableAction::Mate,
        ast::Action::Subcondition(ref sub) => evaluate(me, other, sub),
    }
}

fn eval_value(me: &Creature, other: &Creature, val: &ast::Value) -> usize {
    match *val {
        ast::Value::Literal(x) => x as usize,
        ast::Value::Random => RngState::from_creatures(me, other)
            .rand_range(0, settings::MAX_GENE_VALUE as usize),
        ast::Value::Me(attr) => me.attr(attr),
        ast::Value::Other(attr) => other.attr(attr),
    }
}
